File: other-tests/src/main/java/org/hzt/PrimeNrCalculator.java
package org.hzt;

import org.hzt.utils.Timer;

import java.util.Arrays;
import java.util.stream.IntStream;

import static org.hzt.utils.It.println;

public final class PrimeNrCalculator {

    private PrimeNrCalculator() {
    }

    /**
     *
     *@see <a href="https://www.geeksforgeeks.org/java-program-for-sieve-of-eratosthenes/">Java Program for Sieve of Eratosthenes</a>
     * @param n the upperbound to search for primes
     */
    public static IntStream sieveOfEratosthenes(int n) {
        boolean[] sieve = new boolean[n + 1];
        Arrays.fill(sieve, true);
        int p = 2;
        while (true) {
            final var pSquared = p * p;
            if (pSquared > n) {
                break;
            }
            // If prime[p] is not changed, then it is a prime
            if (sieve[p]) {
                // Update all multiples of p
                for (int i = pSquared; i <= n; i += p) {
                    sieve[i] = false;
                }
            }
            p++;
        }
        // filter out all nrs that are prime
        return IntStream.rangeClosed(2, n)
                .filter(i -> sieve[i]);
    }

    public static void main(String... args) {
        final var n = 1_000_000_000;
        final var timer = Timer.timeAnIntFunction(n, PrimeNrCalculator::sieveOfEratosthenes);
        final var primes = timer.getResult().toArray();
        final int biggest = primes[primes.length - 1];
        println("primes length = " + primes.length);
        println("duration = " + timer.formattedDurationInSeconds());
        println("biggest = " + biggest);
    }
}

File: other-tests/src/main/java/org/hzt/PrimeNrCalculator.kt
package org.hzt

import org.hzt.utils.Timer
import java.util.stream.IntStream

fun intStreamBySieveOfEratosthenes(@Suppress("SameParameterValue") n: Int): IntStream =
    sieveOfEratosthenes(n).let { sieve -> IntStream.rangeClosed(2, n).filter { sieve[it] } }

private tailrec fun sieveOfEratosthenes(
    n: Int,
    sieve: BooleanArray = BooleanArray(n + 1) { true },
    p: Int = 2
): BooleanArray {
    val pSquared = p * p
    if (pSquared !in p..n) return sieve
    // If prime[p] is not changed, then it is a prime
    if (sieve[p]) {
        // Update all multiples of p
        for (i in pSquared..n step p) sieve[i] = false
    }
    return sieveOfEratosthenes(n, sieve, p + 1)
}

fun main() {
    val n = Int.MAX_VALUE - 3
    val timer = Timer.timeAnIntFunction(n, ::intStreamBySieveOfEratosthenes)
    val primes = timer.result.toArray()
    val biggest = primes[primes.size - 1]
    println("primes length = " + primes.size)
    println("duration = " + timer.formattedDurationInSeconds())
    println("biggest = $biggest")
    require(biggest.isPrime())
    require(biggest == 2_147_483_629)
}

fun Int.isPrime(): Boolean = this > 1 && (2..this / 2).none { this % it == 0 }
